pycqed/instrument_drivers/physical_instruments/_QCC/QCCMicrocode.py

```diff
@@ -0,0 +1,284 @@
+        raise ValueError('bin_to_hex: parameter {} is not a number'.format(
+            bin_val))
+
+    return format(int(bin_val, 2), 'X').zfill(hex_width)
+
+
+def check_int(d, left_bound, right_bound):
+    if not isinstance(d, int):
+        raise ValueError("The input parameter {} is not an integer.".format(d))
+
+    if d < left_bound or d > right_bound:
+        raise ValueError("Parameter {} out of range: [{}, {}].".format(
+            d, left_bound, right_bound))
+
+    return True
+
+
+def ToFourBytes(int_data):
+    if not isinstance(int_data, int):
+        raise ValueError(
+            "The input data {} is not an integer".format(int_data))
+
+    return int_data.to_bytes(4, byteorder='little', signed=False)
+
+
+class QCCMicrocode():
+    """
+    This is class is used to implement all functionalities required to generate
+    the microcode stored in the control store of CCLight.
+    """
+
+    def __init__(self):
+        self.microcode = [0]*256
+        self.CS_header = "OpTypeLeft  CW_Left  OpTypeRight  CW_Right  Condition"
+        self.CS_format = "{:5d}       {:3d}      {:6d}     {:6d}     {:5d} "
+        self.type_dict = ["None", "Mw", "Flux", "Meas"]
+        print("initialization finished.")
+
+    def gen_control_store_line(self, condition, op_type_left, cw_left,
+                               op_type_right, cw_right, present_format=False):
+        """
+        This function generates a single entry for the control store.
+        The entry structure is:
+            [cond. exe. bits | left u_instruction | right u_instruction]
+
+        - The Conditional execution bits are 2-bit wide.
+          - 0 means this is an unconditional operation.
+          - 1~3 means both the left and right operation will be conditioned on
+                the three different combinatorial results.
+
+        - The microinstruction has the following structure:
+            [Type | Codeword]
+          - Type: a 2-bit value indicating which kind of control to trigger:
+            - 0 this microinstruction is not a valid one. Ignored by QuMA.
+            - 1 --> AWG-8 for microwave control.
+            - 2 --> AWG-8 for flux control
+            - 3 --> UHFQC for Measurement
+          - Codeword: An 8-bit value.
+            - For microwave control, all 8 bits can be used.
+            - For flux control, only the low three bits are used.
+            - For measurement, this field is ignored by QuMA.
+        """
+        check_int(condition, 0, 3)
+        check_int(op_type_left, 0, 3)
+        check_int(cw_left, 0, 255)
+        check_int(op_type_right, 0, 3)
+        check_int(cw_right, 0, 255)
+
+        final_val = (condition << 20) + (op_type_left << 18) +\
+            (cw_left << 10) + (op_type_right << 8) + cw_right
+
+        # return final_val
+        bin_str = get_bin(final_val, 32)
+        # hex_str = bin_to_hex(bin_str, 8)  # commented as not used
+
+        if present_format is True:
```

Comment:
the `is True` part is not needed. `if present_format:` is also sufficient

Reply:
Redundant. Leaving for now for explicit  condition...